src/main/java/sirius/kernel/xml/Outcall.java

```diff
@@ -88,54 +87,126 @@
      */
-    public Outcall(URL url) throws IOException {
-        checkTimeoutBlacklist(url);
+    public Outcall(URI uri) throws IOException {
+        checkTimeoutBlacklist(uri);
 
-        connection = (HttpURLConnection) url.openConnection();
-        connection.setDoInput(true);
-        connection.setConnectTimeout(DEFAULT_CONNECT_TIMEOUT);
-        connection.setReadTimeout(DEFAULT_READ_TIMEOUT);
+        clientBuilder = HttpClient.newBuilder().connectTimeout(DEFAULT_CONNECT_TIMEOUT);
+        requestBuilder = HttpRequest.newBuilder(uri);
     }
 
-    private void checkTimeoutBlacklist(URL url) throws IOException {
-        if (connectTimeoutBlacklistDuration.isZero()) {
-            return;
+    /**
+     * Creates a new <tt>Outcall</tt> with the given client and request.
+     *
+     * @param client  the http client to use
+     * @param request the request to execute
+     * @throws IOException if the host is blacklisted
+     */
+    public Outcall(HttpClient client, HttpRequest request) throws IOException {
+        checkTimeoutBlacklist(request.uri());
+
+        this.client = client;
+        this.request = request;
+    }
+
+    /**
+     * Allows to modify the client before the request is sent by returning the builder that is used to create it.
+     *
+     * @return the underlying {@link HttpClient.Builder}
+     */
+    public HttpClient.Builder modifyClient() {
+        if (client != null) {
+            throw new IllegalStateException("Can no longer modify client, request has already been sent!");
         }
+        return clientBuilder;
+    }
 
-        Long timeout = timeoutBlacklist.get(url.getHost());
-        if (timeout != null) {
-            if (timeout > System.currentTimeMillis()) {
-                throw new IOException(Strings.apply(
-                        "Connecting to host %s is currently rejected due to connectivity issues.",
-                        url.getHost()));
-            } else {
-                timeoutBlacklist.remove(url.getHost());
-            }
+    /**
+     * Allows to modify the request before the request is sent by returning the builder that is used to create it.
+     *
+     * @return the underlying {@link HttpRequest.Builder}
+     */
+    public HttpRequest.Builder modifyRequest() {
+        if (request != null) {
+            throw new IllegalStateException("Can no longer modify request, request has already been sent!");
         }
+        return requestBuilder;
+    }
+
+    /**
+     * Executes the outcall with the given {@link java.net.http.HttpResponse.BodyHandler response handler}.
+     *
+     * @param handler a response body handler
+     * @param <R>     the body type
+     * @return the response, typed by the given handler
+     * @throws IOException in case of any IO error
+     * @see java.net.http.HttpResponse.BodyHandlers
+     */
+    @SuppressWarnings("unchecked")
+    public <R> HttpResponse<R> doCall(HttpResponse.BodyHandler<R> handler) throws IOException {
```

Comment:
do we need this and all the handlers below? is there any benefit for introducing the additional complexity?